internal/controller/olsconfig_controller.go

```diff
@@ -429,27 +546,141 @@ func (r *OLSConfigReconciler) checkDeploymentStatus(deployment *appsv1.Deploymen
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil // Resource will be picked up on next reconciliation
+		}
+		return err
+	}
+
+	if secret.Annotations == nil {
+		secret.Annotations = make(map[string]string)
+	}
+
+	if _, exists := secret.Annotations[utils.WatcherAnnotationKey]; exists {
+		return nil // Already annotated
+	}
+
+	secret.Annotations[utils.WatcherAnnotationKey] = utils.OLSConfigName
+	return r.Update(ctx, secret)
+}
+
+// annotateConfigMapIfNeeded annotates a configmap with the watcher annotation if it doesn't already have it.
+// Returns nil if the configmap doesn't exist (will be picked up on next reconciliation).
+func (r *OLSConfigReconciler) annotateConfigMapIfNeeded(ctx context.Context, name, namespace string) error {
+	cm := &corev1.ConfigMap{}
+	err := r.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, cm)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil // Resource will be picked up on next reconciliation
+		}
+		return err
+	}
+
+	if cm.Annotations == nil {
+		cm.Annotations = make(map[string]string)
+	}
+
+	if _, exists := cm.Annotations[utils.WatcherAnnotationKey]; exists {
+		return nil // Already annotated
+	}
+
+	cm.Annotations[utils.WatcherAnnotationKey] = utils.OLSConfigName
+	return r.Update(ctx, cm)
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *OLSConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	r.Logger = ctrl.Log.WithName("Reconciler")
-	r.StateCache = make(map[string]string)
 	r.NextReconcileTime = time.Now()
 
-	generationChanged := builder.WithPredicates(predicate.GenerationChangedPredicate{})
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&olsv1alpha1.OLSConfig{}).
-		Owns(&appsv1.Deployment{}, generationChanged).
+		Owns(&appsv1.Deployment{}).
 		Owns(&corev1.ServiceAccount{}).
 		Owns(&rbacv1.ClusterRole{}).
 		Owns(&rbacv1.ClusterRoleBinding{}).
 		Owns(&corev1.Service{}).
 		Owns(&corev1.ConfigMap{}).
 		Owns(&corev1.Secret{}).
 		Owns(&corev1.PersistentVolumeClaim{}).
-		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(watchers.SecretWatcherFilter)).
-		Watches(&corev1.ConfigMap{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
-			return watchers.ConfigMapWatcherFilter(r, ctx, obj, r.Options.UseLCore)
-		})).
+		Watches(&corev1.Secret{},
+			&watchers.SecretUpdateHandler{Reconciler: r},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					// For Create events, allow all secrets in our namespace
+					// This handles recreated secrets that don't have annotations yet
+					return e.Object.GetNamespace() == r.Options.Namespace
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					// For Update events, use strict filtering
+					return r.shouldWatchSecret(e.ObjectNew)
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
```

Comment:
I think we can reconcile for delete of Secrets in our namespace @blublinsky

Reply:
delete does not recomcile, create does.

Reply:
We can, but why? If the user deletes the secret everything will continue to work, as we are copying it. We update on creation and/or update